### PR TITLE
feat: deeper intellisense support

### PR DESCRIPTION
**Pull request by jd-carroll**
Background: I am new to Tailwind (and still pretty green with CSS generally). The IntelliSense and linting from the Tailwind VS Code extension has been my saving grace to understand what is going on.

However, out of the box the extension has pretty limited (or focused) areas where the IntelliSense will kick in.

This change expands the number of places where the Tailwind extension will provide IntelliSense support.

Any object with `Styles` (case sensitive) in the name will now trigger IntelliSense.  This is particularly useful in places like `Button`. Previously, no IntelliSense was provided to the `buttonVariant` styles.

**With this change (and some moving around), all `buttonVariant` styles have IntelliSense support.**

Example:
image

Note: The core change is to add `".*Styles*"` to the list of `tailwindCSS.classAttributes` which is where Tailwind knows to look in the file for IntelliSense. This was added to the root `settings.json` as well as the `settings.json` under the template.

### CODE REVIEW

**Comment by ap-1** [comment]
Relevant: CVA's [tailwindCSS.experimental.classRegex](https://cva.style/docs/installation#intellisense)

**Review by shadcn — CHANGES_REQUESTED** [review]
This is great. I have one minor comment re. the button default styles.

**Comment by shadcn** [comment]
@jd-carroll What do you think of @ap-1 PR here https://github.com/shadcn/ui/pull/439? This adds cva support but it's not limited to *Styles and would work with any cva setup.

**Comment by ap-1** [comment]
> @jd-carroll What do you think of @ap-1 PR here #439? This adds cva support but it's not limited to *Styles and would work with any cva setup.

I just expanded my PR to also target classes inside of the `cn()` helper

**Comment by jd-carroll** [comment]
So the biggest difference between `tailwindCSS.classAttributes` and `tailwindCSS.experimental.classRegex` is that the `classRegex` variant is autocomplete only. Meaning, if you tried to hover over any of those class names inside of `cn` or `cva` you **will not** receive any feedback.

For me the IntelliSense on hover is the most frequently used part of my workflow.

Further, if you are going to focus on adding auto-complete, you probably need to set this setting too:
```
"editor.quickSuggestions": {
  "strings": true
}
```

By default, VS Code will not attempt to auto-complete strings unless you prompt it. Enabling the `quickSuggestions` will tell VS Code to always provide auto-complete suggestions (for all strings). -> For me that is not desired

Lastly, there is a difference in performance in how Tailwind "finds" each section.  I am not familiar with the details or why (presumably using `classAttributes` does not entail RegExp), but some of the Tailwind IntelliSense documentation seemed to hint at performance impacts if you have too many `classRegex`'s defined.

I even tried the `classRegex` approach with this regex:
```json
  "tailwindCSS.experimental.classRegex": [
    [
      "/\\*tw\\*/ ([^;]*);", "['\"` \\t\\n\\r\\f\\v]*([^'\"` \\t\\n\\r\\f\\v]*)['\"` \\t\\n\\r\\f\\v]"
    ]
  ],
```

In summary, it will look for any class names with a leading `/*tw*/`.

**However, at the end of the day, I don't think it matters. Using `classAttributes` is my prefered approach because of the hover support. But I think it is perfectly rational to have both if desired.**

**Comment by ap-1** [comment]
> So the biggest difference between `tailwindCSS.classAttributes` and `tailwindCSS.experimental.classRegex` is that the `classRegex` variant is autocomplete only. Meaning, if you tried to hover over any of those class names inside of `cn` or `cva` you **will not** receive any feedback.

image

I haven't created a minimal repro, but I can hover `cva` class names in [one of my repos](https://github.com/ap-1/balladeer). As expected, I only have the ability to do this when I add `tailwindCSS.experimental.classRegex` snippet:

```json
"tailwindCSS.experimental.classRegex": [
  ["cva\\(([^)]*)\\)", "[\"'`]([^\"'`]*).*?[\"'`]"]
],
``` 

image

When the snippet is present, I also have autocomplete despite having the following setting:

```json
"editor.quickSuggestions": {
  "strings": false
}
```

**Comment by jd-carroll** [comment]
Fascinating, then I must have been doing something wrong…

In that case I see no difference between the two, and it is just a
judgement call.

It may make sense to use the classRegex approach as it would seem like it
slots into the existing code base the best

**Comment by jd-carroll** [comment]
Closing as I think @ap-1's approach is less intrusive for the codebase.

**Comment by ap-1** [comment]
Since this was closed, I moved the Tailwind CSS Extension docs over to my PR.
